qiskit/algorithms/evolvers/evolution_problem.py

```diff
@@ -44,15 +45,117 @@ def __init__(
+
+        Raises:
+            ValueError: If no Hamiltonian is provided.
+        """
+        if hamiltonian is None:
+            raise ValueError(
+                "No ``hamiltonian`` provided for the EvolutionProblem. It is required."
+            )
+
+        self._check_parameters(hamiltonian, self.hamiltonian_value_dict, self.t_param)
+        self._hamiltonian = hamiltonian
+
+    @property
+    def time(self) -> float:
+        """Returns time."""
+        return self._time
+
+    @time.setter
+    def time(self, time: float) -> None:
+        """
+        Sets time and validates it.
+
+        Raises:
+            ValueError: If time is not positive.
+        """
+        if time <= 0:
+            raise ValueError(f"Evolution time must be > 0 but was {time}.")
+        self._time = time
+
+    @property
+    def initial_state(self) -> Union[StateFn, QuantumCircuit]:
+        """Returns an initial state."""
+        return self._initial_state
+
+    @initial_state.setter
+    def initial_state(self, initial_state: Union[StateFn, QuantumCircuit]) -> None:
+        """
+        Sets an initial state and validates it.
+
+        Raises:
+            ValueError: If no initial state is provided.
+        """
+        if initial_state is None:
+            raise ValueError(
+                "No ``initial_state`` provided for the EvolutionProblem. It is required."
+            )
+        self._initial_state = initial_state
+
+    def _check_parameters(
+        self,
+        hamiltonian: Union[OperatorBase, BaseOperator],
+        hamiltonian_value_dict: Optional[Dict[Parameter, Union[complex]]] = None,
+        t_param: Optional[Parameter] = None,
+    ) -> None:
+        """
+        Checks if all parameters present in the Hamiltonian are also present in the dictionary
+        that maps them to values.
+        Args:
+            hamiltonian: The Hamiltonian under which to evolve the system.
+            hamiltonian_value_dict: If the Hamiltonian contains free parameters, this
+                dictionary maps all these parameters to values.
+            t_param: Time parameter in case of a time-dependent Hamiltonian. This
+            free parameter must be within the ``hamiltonian``.
+
+        Raises:
+            ValueError: If Hamiltonian parameters cannot be bound with data provided.
+        """
+        if isinstance(hamiltonian, OperatorBase):
+            t_param_set = set()
+            if t_param is not None:
+                t_param_set.add(t_param)
+            hamiltonian_dict_param_set = set()
+            if hamiltonian_value_dict is not None:
+                hamiltonian_dict_param_set = hamiltonian_dict_param_set.union(
+                    set(hamiltonian_value_dict.keys())
+                )
+            params_set = t_param_set.union(hamiltonian_dict_param_set)
+            hamiltonian_param_set = set(hamiltonian.parameters)
+
+            if hamiltonian_param_set != params_set:
```

Comment:
You want to enforce strict equality here, e..g if I provided more parameter/value keys pairs than were needed, where the hamiltonian params were a subset of what was provided, this we still want to fail on. I cannot though see a good reason why the user would have a larger set if params, where this was part thereof so I guess its all good.

Reply:
Yes, I think it is good to be strict here. Having some additional parameters in the dict more likely indicates some discrepancy (error) in input than a valid scenario in my opinion.